server/src/main/java/io/unitycatalog/server/service/Scim2UserService.java

```diff
@@ -0,0 +1,204 @@
+            .filter(m -> match(filterEvaluator, userFilter, m))
+            .skip(startIndex.orElse(1) - 1)
+            .limit(count.orElse(50))
+            .toList();
+    return HttpResponse.ofJson(userResourcesList);
+  }
+
+  @Post("")
+  public HttpResponse createScimUser(UserResource userResource) {
+    // Get primary email address
+    Email primaryEmail =
+        userResource.getEmails().stream()
+            .filter(Email::getPrimary)
+            .findFirst()
+            .orElseThrow(
+                () ->
+                    new Scim2RuntimeException(
+                        new PreconditionFailedException("User does not have a primary email.")));
+
+    try {
+      User user = USER_REPOSITORY.getUserByEmail(primaryEmail.getValue());
+      return HttpResponse.ofJson(asUserResource(user));
+    } catch (BaseException e) {
+      if (e.getErrorCode() == ErrorCode.NOT_FOUND) {
+        String pictureUrl = "";
+        if (userResource.getPhotos() != null && !userResource.getPhotos().isEmpty()) {
+          pictureUrl = userResource.getPhotos().get(0).getValue().toString();
+        }
+        User user =
+            USER_REPOSITORY.createUser(
+                new CreateUser()
+                    .name(userResource.getDisplayName())
+                    .email(primaryEmail.getValue())
+                    .externalId(userResource.getExternalId())
+                    .pictureUrl(pictureUrl));
+        return HttpResponse.ofJson(asUserResource(user));
+      } else {
+        throw e;
+      }
+    }
+  }
+
+  @Get("/self")
+  public HttpResponse getCurrentUser() {
+    ServiceRequestContext ctx = ServiceRequestContext.current();
+    DecodedJWT decodedJWT = ctx.attr(AuthDecorator.DECODED_JWT_ATTR);
+    Claim sub = decodedJWT.getClaim("sub");
+    return HttpResponse.ofJson(asUserResource(USER_REPOSITORY.getUserByEmail(sub.asString())));
+  }
+
+  @Get("/{id}")
+  public HttpResponse getUser(@Param("id") String id) {
+    return HttpResponse.ofJson(asUserResource(USER_REPOSITORY.getUser(id)));
+  }
+
+  @Put("/{id}")
+  public HttpResponse updateUser(@Param("id") String id, UserResource userResource) {
+    UserResource user = asUserResource(USER_REPOSITORY.getUser(id));
+    if (!id.equals(userResource.getId())) {
+      throw new Scim2RuntimeException(new ResourceConflictException("User id mismatch."));
+    }
+    // Get primary email address
+    Email primaryEmail =
+        userResource.getEmails().stream()
+            .filter(Email::getPrimary)
+            .findFirst()
+            .orElse(user.getEmails().get(0));
+
+    String newName = userResource.getDisplayName();
+    if (newName == null) {
+      newName = user.getDisplayName();
+    }
+
+    String externalId = userResource.getExternalId();
+    if (externalId == null) {
+      externalId = user.getExternalId();
+    }
+
+    UpdateUser updateUser =
+        new UpdateUser().newName(newName).email(primaryEmail.getValue()).externalId(externalId);
```

Comment:
Similar to `CreateUser`, `UpdateUser` is also used only within the server code, so is it necessary to define it in the yaml spec?

Reply:
cc: @xx-db

Reply:
Done.